{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas \n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#FILE</th>\n",
       "      <th>SEQUENCE</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>GENE</th>\n",
       "      <th>COVERAGE</th>\n",
       "      <th>COVERAGE_MAP</th>\n",
       "      <th>GAPS</th>\n",
       "      <th>%COVERAGE</th>\n",
       "      <th>%IDENTITY</th>\n",
       "      <th>DATABASE</th>\n",
       "      <th>ACCESSION</th>\n",
       "      <th>PRODUCT</th>\n",
       "      <th>RESISTANCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AA010_contigs.fa</td>\n",
       "      <td>contig00001</td>\n",
       "      <td>113766</td>\n",
       "      <td>114708</td>\n",
       "      <td>-</td>\n",
       "      <td>VFG037748(gb|YP_001085628)</td>\n",
       "      <td>1-943/984</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>95.83</td>\n",
       "      <td>82.72</td>\n",
       "      <td>VFDB_full</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VFG037748(gb|YP_001085628) (lpxM) lipid A bios...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AA010_contigs.fa</td>\n",
       "      <td>contig00001</td>\n",
       "      <td>380608</td>\n",
       "      <td>382042</td>\n",
       "      <td>-</td>\n",
       "      <td>VFG037735(gb|WP_000633124)</td>\n",
       "      <td>6-1440/1449</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>99.03</td>\n",
       "      <td>84.18</td>\n",
       "      <td>VFDB_full</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VFG037735(gb|WP_000633124) (adeH) outer membra...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AA010_contigs.fa</td>\n",
       "      <td>contig00001</td>\n",
       "      <td>382060</td>\n",
       "      <td>385239</td>\n",
       "      <td>-</td>\n",
       "      <td>VFG037724(gb|WP_001027061)</td>\n",
       "      <td>1-3180/3180</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>91.26</td>\n",
       "      <td>VFDB_full</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VFG037724(gb|WP_001027061) (adeG) cation/multi...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AA010_contigs.fa</td>\n",
       "      <td>contig00001</td>\n",
       "      <td>385246</td>\n",
       "      <td>386466</td>\n",
       "      <td>-</td>\n",
       "      <td>VFG037705(gb|WP_000010636)</td>\n",
       "      <td>1-1221/1221</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>87.31</td>\n",
       "      <td>VFDB_full</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VFG037705(gb|WP_000010636) (adeF) membrane-fus...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AA010_contigs.fa</td>\n",
       "      <td>contig00001</td>\n",
       "      <td>399065</td>\n",
       "      <td>400822</td>\n",
       "      <td>+</td>\n",
       "      <td>VFG050781(gb|WP_001020905.1)</td>\n",
       "      <td>1-1758/1758</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>85.67</td>\n",
       "      <td>VFDB_full</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VFG050781(gb|WP_001020905.1) (gspE2) putative ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              #FILE     SEQUENCE   START     END STRAND  \\\n",
       "0  AA010_contigs.fa  contig00001  113766  114708      -   \n",
       "1  AA010_contigs.fa  contig00001  380608  382042      -   \n",
       "2  AA010_contigs.fa  contig00001  382060  385239      -   \n",
       "3  AA010_contigs.fa  contig00001  385246  386466      -   \n",
       "4  AA010_contigs.fa  contig00001  399065  400822      +   \n",
       "\n",
       "                           GENE     COVERAGE     COVERAGE_MAP GAPS  %COVERAGE  \\\n",
       "0    VFG037748(gb|YP_001085628)    1-943/984  ===============  0/0      95.83   \n",
       "1    VFG037735(gb|WP_000633124)  6-1440/1449  ===============  0/0      99.03   \n",
       "2    VFG037724(gb|WP_001027061)  1-3180/3180  ===============  0/0     100.00   \n",
       "3    VFG037705(gb|WP_000010636)  1-1221/1221  ===============  0/0     100.00   \n",
       "4  VFG050781(gb|WP_001020905.1)  1-1758/1758  ===============  0/0     100.00   \n",
       "\n",
       "   %IDENTITY   DATABASE  ACCESSION  \\\n",
       "0      82.72  VFDB_full        NaN   \n",
       "1      84.18  VFDB_full        NaN   \n",
       "2      91.26  VFDB_full        NaN   \n",
       "3      87.31  VFDB_full        NaN   \n",
       "4      85.67  VFDB_full        NaN   \n",
       "\n",
       "                                             PRODUCT  RESISTANCE  \n",
       "0  VFG037748(gb|YP_001085628) (lpxM) lipid A bios...         NaN  \n",
       "1  VFG037735(gb|WP_000633124) (adeH) outer membra...         NaN  \n",
       "2  VFG037724(gb|WP_001027061) (adeG) cation/multi...         NaN  \n",
       "3  VFG037705(gb|WP_000010636) (adeF) membrane-fus...         NaN  \n",
       "4  VFG050781(gb|WP_001020905.1) (gspE2) putative ...         NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import data \n",
    "MosAIC = pd.read_csv(\"September_MosAIC_LSTM_UW_Metadata_LB2.csv\")\n",
    "MosAIC.head()\n",
    "\n",
    "dRepClusters = pd.read_csv(\"Cdb.csv\")\n",
    "dRepClusters.head()\n",
    "\n",
    "VF_Genes = pd.read_csv(\"210423_MosAIC_VF_Search.tsv\", delimiter=\"\\t\")\n",
    "VF_Genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    internal_id seq_run seq_sumission_date data_return_date  \\\n",
      "0       USHLN06    UW-1          29-Sep-21        01-Nov-21   \n",
      "1       USHLN08    UW-1          29-Sep-21        01-Nov-21   \n",
      "2      USHLN005    UW-2          07-Jul-22        19-Aug-22   \n",
      "3       USHLN16    UW-1          29-Sep-21        01-Nov-21   \n",
      "4      USHLN010    UW-3          23-Aug-22        09-Sep-22   \n",
      "..          ...     ...                ...              ...   \n",
      "413     USHLN33    UW-1          29-Sep-21        01-Nov-21   \n",
      "415     USHLN34    UW-1          29-Sep-21        01-Nov-21   \n",
      "416     USHLN35    UW-1          29-Sep-21        01-Nov-21   \n",
      "417    USHLN028    UW-2          07-Jul-22        19-Aug-22   \n",
      "418    USHLN029    UW-2          07-Jul-22        19-Aug-22   \n",
      "\n",
      "    internal_date_isolate_received     source_lab isolation_date  \\\n",
      "0                        06-May-21  Brackney_Doug      08-Jun-18   \n",
      "1                        06-May-21  Brackney_Doug      13-Jun-18   \n",
      "2                        06-May-21  Brackney_Doug      16-Jun-18   \n",
      "3                        06-May-21  Brackney_Doug      17-Jun-18   \n",
      "4                        06-May-21  Brackney_Doug      17-Jun-18   \n",
      "..                             ...            ...            ...   \n",
      "413                      13-Jul-21    Xu_Jiannong         Jul-16   \n",
      "415                      13-Jul-21    Xu_Jiannong         Jul-16   \n",
      "416                      13-Jul-21    Xu_Jiannong         Jul-16   \n",
      "417                      13-Jul-21    Xu_Jiannong         Jul-16   \n",
      "418                      13-Jul-21    Xu_Jiannong         Jul-16   \n",
      "\n",
      "                       sample_location  \\\n",
      "0       New_Haven,CT,USA;North_America   \n",
      "1       New_Haven,CT,USA;North_America   \n",
      "2       New_Haven,CT,USA;North_America   \n",
      "3       New_Haven,CT,USA;North_America   \n",
      "4       New_Haven,CT,USA;North_America   \n",
      "..                                 ...   \n",
      "413  St_Augustine,FL,USA;North_America   \n",
      "415  St_Augustine,FL,USA;North_America   \n",
      "416  St_Augustine,FL,USA;North_America   \n",
      "417  St_Augustine,FL,USA;North_America   \n",
      "418  St_Augustine,FL,USA;North_America   \n",
      "\n",
      "                                   sample_location_lat sample_location_lon  \\\n",
      "0                                            41.316307          -72.922585   \n",
      "1                                            41.316307          -72.922585   \n",
      "2                                            41.316307          -72.922585   \n",
      "3                                            41.316307          -72.922585   \n",
      "4                                            41.316307          -72.922585   \n",
      "..                                                 ...                 ...   \n",
      "413                                          29.895568             -81.313   \n",
      "415                                          29.895568             -81.313   \n",
      "416                                          29.895568             -81.313   \n",
      "417  Collected from one of six sites (lat, lon): (1...                 NaN   \n",
      "418  Collected from one of six sites (lat, lon): (1...                 NaN   \n",
      "\n",
      "     ...               gtdb_species fastani_reference fastani_ani fastani_af  \\\n",
      "0    ...       Klebsiella variicola   GCF_000828055.2       99.26       0.95   \n",
      "1    ...             Serratia nevei   GCF_016742975.1       98.81       0.88   \n",
      "2    ...        Cedecea sp000277545   GCF_000277545.1       99.98       0.97   \n",
      "3    ...   Klebsiella michiganensis   GCF_002925905.1       98.52       0.89   \n",
      "4    ...         Rouxiella badensis   GCF_002093665.1       99.65       0.98   \n",
      "..   ...                        ...               ...         ...        ...   \n",
      "413  ...         Erwinia aphidicola   GCF_014773485.1       99.43       0.92   \n",
      "415  ...           Pantoea ananatis   GCF_000710035.2       98.88       0.92   \n",
      "416  ...           Pantoea ananatis   GCF_000710035.2       99.02       0.91   \n",
      "417  ...  Lysinibacillus fusiformis   GCF_003049525.1       97.20       0.93   \n",
      "418  ...  Lysinibacillus fusiformis   GCF_003049525.1       96.85       0.92   \n",
      "\n",
      "    SR_accession_no ER_accession_no fastani_ani.1 fastani_af.1  \\\n",
      "0               NaN             NaN         99.26         0.95   \n",
      "1               NaN             NaN         98.81         0.88   \n",
      "2               NaN             NaN         99.98         0.97   \n",
      "3               NaN             NaN         98.52         0.89   \n",
      "4               NaN             NaN         99.65         0.98   \n",
      "..              ...             ...           ...          ...   \n",
      "413             NaN             NaN         99.43         0.92   \n",
      "415             NaN             NaN         98.88         0.92   \n",
      "416             NaN             NaN         99.02         0.91   \n",
      "417             NaN             NaN         97.20         0.93   \n",
      "418             NaN             NaN         96.85         0.92   \n",
      "\n",
      "    SR_accession_no.1 ER_accession_no.1  \n",
      "0                 NaN               NaN  \n",
      "1                 NaN               NaN  \n",
      "2                 NaN               NaN  \n",
      "3                 NaN               NaN  \n",
      "4                 NaN               NaN  \n",
      "..                ...               ...  \n",
      "413               NaN               NaN  \n",
      "415               NaN               NaN  \n",
      "416               NaN               NaN  \n",
      "417               NaN               NaN  \n",
      "418               NaN               NaN  \n",
      "\n",
      "[392 rows x 52 columns]\n"
     ]
    }
   ],
   "source": [
    "# filter MosAIC df with genomes that have passed QC \n",
    "MosAIC_pass = MosAIC[MosAIC['qc_pass'] == 'Pass']\n",
    "print(MosAIC_pass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   #FILE     SEQUENCE   START     END STRAND                          GENE  \\\n",
      "0  AA010  contig00001  113766  114708      -    VFG037748(gb|YP_001085628)   \n",
      "1  AA010  contig00001  380608  382042      -    VFG037735(gb|WP_000633124)   \n",
      "2  AA010  contig00001  382060  385239      -    VFG037724(gb|WP_001027061)   \n",
      "3  AA010  contig00001  385246  386466      -    VFG037705(gb|WP_000010636)   \n",
      "4  AA010  contig00001  399065  400822      +  VFG050781(gb|WP_001020905.1)   \n",
      "\n",
      "      COVERAGE     COVERAGE_MAP GAPS  %COVERAGE  %IDENTITY   DATABASE  \\\n",
      "0    1-943/984  ===============  0/0      95.83      82.72  VFDB_full   \n",
      "1  6-1440/1449  ===============  0/0      99.03      84.18  VFDB_full   \n",
      "2  1-3180/3180  ===============  0/0     100.00      91.26  VFDB_full   \n",
      "3  1-1221/1221  ===============  0/0     100.00      87.31  VFDB_full   \n",
      "4  1-1758/1758  ===============  0/0     100.00      85.67  VFDB_full   \n",
      "\n",
      "   ACCESSION                                            PRODUCT  RESISTANCE  \n",
      "0        NaN  VFG037748(gb|YP_001085628) (lpxM) lipid A bios...         NaN  \n",
      "1        NaN  VFG037735(gb|WP_000633124) (adeH) outer membra...         NaN  \n",
      "2        NaN  VFG037724(gb|WP_001027061) (adeG) cation/multi...         NaN  \n",
      "3        NaN  VFG037705(gb|WP_000010636) (adeF) membrane-fus...         NaN  \n",
      "4        NaN  VFG050781(gb|WP_001020905.1) (gspE2) putative ...         NaN  \n",
      "\n",
      "     genome secondary_cluster  threshold cluster_method comparison_algorithm  \\\n",
      "0  MMO-105               1_1       0.05        average              fastANI   \n",
      "1   MMO-16               1_1       0.05        average              fastANI   \n",
      "2   MMO-21               1_1       0.05        average              fastANI   \n",
      "3  USMM094               1_1       0.05        average              fastANI   \n",
      "4  USMM095               1_1       0.05        average              fastANI   \n",
      "\n",
      "   primary_cluster  \n",
      "0                1  \n",
      "1                1  \n",
      "2                1  \n",
      "3                1  \n",
      "4                1  \n"
     ]
    }
   ],
   "source": [
    "# function to remove _contigs.fa extension \n",
    "\n",
    "def remove_file_extension(df, column_name):\n",
    "    '''\n",
    "    df = you dataframe\n",
    "    column_name = target column name in ' '\n",
    "    '''\n",
    "    for column in [column_name]: \n",
    "        df[column_name] = df[column_name].str.replace('_contigs.fa', '')\n",
    "\n",
    "    return(df)\n",
    "\n",
    "VF_Genes_Clean = remove_file_extension(VF_Genes, '#FILE')\n",
    "dRep_Clusters_Clean = remove_file_extension(dRepClusters, 'genome')\n",
    "\n",
    "print(VF_Genes_Clean.head())\n",
    "print(\"\\n\", dRep_Clusters_Clean.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'MMO-59', 'MMO-93'}\n",
      "set()\n"
     ]
    }
   ],
   "source": [
    "# Evaluate file deliniators\n",
    "'''\n",
    "compare the ID names\n",
    "'''\n",
    "set1 = set(MosAIC_pass['internal_id'])\n",
    "set2 = set(dRepClusters['genome'])\n",
    "set3 = set(VF_Genes_Clean['#FILE'])\n",
    "'''\n",
    "file name has _contigs.fa\n",
    "'''\n",
    "\n",
    "print(set2.difference(set1))\n",
    "print(set3.difference(set1))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['internal_id', 'seq_run', 'seq_sumission_date', 'data_return_date',\n",
      "       'internal_date_isolate_received', 'source_lab', 'isolation_date',\n",
      "       'sample_location', 'sample_location_lat', 'sample_location_lon',\n",
      "       'sample_source', 'lab_field_derived', 'host_species_non_mosquito',\n",
      "       'life_stage_non_mosquito', 'sex_non_mosquito', 'tissue_non_mosquito',\n",
      "       'mosquito_species', 'mosquito_strain_lab', 'mosquito_life_stage',\n",
      "       'mosquito_sex', 'mosquito_female_feeding_status', 'mosquito_tissue',\n",
      "       'confirmed_growth_media', 'confirmed_growth_temp', 'qc_pass',\n",
      "       'completeness', 'contamination', 'genome_size_bp', 'number_contigs',\n",
      "       'mean_contig_length_bp', 'longest_contig_bp', 'gc',\n",
      "       'number_predicted_genes', 'n50', 'total_reads', 'avg_coverage',\n",
      "       'gtdb_domain', 'gtdb_parent', 'gtbd_class', 'gtdb_order', 'gtdb_family',\n",
      "       'gtdb_genus', 'gtdb_species', 'fastani_reference', 'fastani_ani',\n",
      "       'fastani_af', 'SR_accession_no', 'ER_accession_no', 'fastani_ani.1',\n",
      "       'fastani_af.1', 'SR_accession_no.1', 'ER_accession_no.1', '#FILE',\n",
      "       'SEQUENCE', 'START', 'END', 'STRAND', 'GENE', 'COVERAGE',\n",
      "       'COVERAGE_MAP', 'GAPS', '%COVERAGE', '%IDENTITY', 'DATABASE',\n",
      "       'ACCESSION', 'PRODUCT', 'RESISTANCE'],\n",
      "      dtype='object')\n",
      "      internal_id     source_lab sample_source lab_field_derived  \\\n",
      "0         USHLN06  Brackney_Doug      mosquito               lab   \n",
      "1         USHLN06  Brackney_Doug      mosquito               lab   \n",
      "2         USHLN06  Brackney_Doug      mosquito               lab   \n",
      "3         USHLN06  Brackney_Doug      mosquito               lab   \n",
      "4         USHLN06  Brackney_Doug      mosquito               lab   \n",
      "...           ...            ...           ...               ...   \n",
      "11875    USHLN029    Xu_Jiannong      mosquito             field   \n",
      "11876    USHLN029    Xu_Jiannong      mosquito             field   \n",
      "11877    USHLN029    Xu_Jiannong      mosquito             field   \n",
      "11878    USHLN029    Xu_Jiannong      mosquito             field   \n",
      "11879    USHLN029    Xu_Jiannong      mosquito             field   \n",
      "\n",
      "      mosquito_tissue  mosquito_species mosquito_female_feeding_status  \\\n",
      "0              midgut     Aedes_aegypti                      blood_fed   \n",
      "1              midgut     Aedes_aegypti                      blood_fed   \n",
      "2              midgut     Aedes_aegypti                      blood_fed   \n",
      "3              midgut     Aedes_aegypti                      blood_fed   \n",
      "4              midgut     Aedes_aegypti                      blood_fed   \n",
      "...               ...               ...                            ...   \n",
      "11875      whole_body  Aedes_albopictus                  non_blood_fed   \n",
      "11876      whole_body  Aedes_albopictus                  non_blood_fed   \n",
      "11877      whole_body  Aedes_albopictus                  non_blood_fed   \n",
      "11878      whole_body  Aedes_albopictus                  non_blood_fed   \n",
      "11879      whole_body  Aedes_albopictus                  non_blood_fed   \n",
      "\n",
      "      mosquito_sex      gtdb_genus               gtdb_species  \\\n",
      "0           female      Klebsiella       Klebsiella variicola   \n",
      "1           female      Klebsiella       Klebsiella variicola   \n",
      "2           female      Klebsiella       Klebsiella variicola   \n",
      "3           female      Klebsiella       Klebsiella variicola   \n",
      "4           female      Klebsiella       Klebsiella variicola   \n",
      "...            ...             ...                        ...   \n",
      "11875       female  Lysinibacillus  Lysinibacillus fusiformis   \n",
      "11876       female  Lysinibacillus  Lysinibacillus fusiformis   \n",
      "11877       female  Lysinibacillus  Lysinibacillus fusiformis   \n",
      "11878       female  Lysinibacillus  Lysinibacillus fusiformis   \n",
      "11879       female  Lysinibacillus  Lysinibacillus fusiformis   \n",
      "\n",
      "                             GENE  ACCESSION  \\\n",
      "0      VFG048533(gb|WP_012541740)        NaN   \n",
      "1      VFG048692(gb|WP_012541980)        NaN   \n",
      "2      VFG048689(gb|WP_002902160)        NaN   \n",
      "3      VFG048671(gb|WP_012541981)        NaN   \n",
      "4      VFG048667(gb|WP_012968412)        NaN   \n",
      "...                           ...        ...   \n",
      "11875  VFG043991(gb|WP_031416422)        NaN   \n",
      "11876  VFG044222(gb|WP_000753441)        NaN   \n",
      "11877  VFG044221(gb|WP_000042061)        NaN   \n",
      "11878  VFG044220(gb|WP_000631132)        NaN   \n",
      "11879  VFG044219(gb|WP_000590068)        NaN   \n",
      "\n",
      "                                                 PRODUCT  RESISTANCE  \n",
      "0      VFG048533(gb|WP_012541740) (iroE) siderophore ...         NaN  \n",
      "1      VFG048692(gb|WP_012541980) (clpV/tssH) type VI...         NaN  \n",
      "2      VFG048689(gb|WP_002902160) (hcp/tssD) type VI ...         NaN  \n",
      "3      VFG048671(gb|WP_012541981) (ompA) OmpA family ...         NaN  \n",
      "4      VFG048667(gb|WP_012968412) (dotU/tssL) type VI...         NaN  \n",
      "...                                                  ...         ...  \n",
      "11875  VFG043991(gb|WP_031416422) (BSPH_RS19015) thio...         NaN  \n",
      "11876  VFG044222(gb|WP_000753441) (GBAA_RS26000) side...         NaN  \n",
      "11877  VFG044221(gb|WP_000042061) (GBAA_RS25995) iron...         NaN  \n",
      "11878  VFG044220(gb|WP_000631132) (GBAA_RS25990) iron...         NaN  \n",
      "11879  VFG044219(gb|WP_000590068) (GBAA_RS25985) side...         NaN  \n",
      "\n",
      "[11880 rows x 14 columns]\n"
     ]
    }
   ],
   "source": [
    "# join data \n",
    "MosAIC_and_VirulenceFactors = pd.merge(MosAIC_pass, VF_Genes_Clean, left_on='internal_id', right_on='#FILE', how = 'left')\n",
    "print(MosAIC_and_VirulenceFactors.columns)\n",
    "\n",
    "# select specific columns\n",
    "MosAIC_VF_select = MosAIC_and_VirulenceFactors[['internal_id', 'source_lab', 'sample_source', 'lab_field_derived', \n",
    "                             'mosquito_tissue', 'mosquito_species','mosquito_female_feeding_status',\n",
    "                             'mosquito_sex', 'gtdb_genus', 'gtdb_species', 'GENE', 'ACCESSION', 'PRODUCT', 'RESISTANCE']]\n",
    "\n",
    "print(MosAIC_VF_select)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>PRODUCT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>internal_id</th>\n",
       "      <th>mosquito_tissue</th>\n",
       "      <th>sample_source</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Achromobacter ruhlandii</th>\n",
       "      <th>MMO-104</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMO-123</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMO-124</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMO-144</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMO-82</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Stenotrophomonas maltophilia</th>\n",
       "      <th>MMO-2</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMO-6</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Stenotrophomonas maltophilia_A</th>\n",
       "      <th>USHLN265</th>\n",
       "      <th>midgut</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>USMM131</th>\n",
       "      <th>midgut</th>\n",
       "      <th>mosquito</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBA7405 sp000755535</th>\n",
       "      <th>AS183</th>\n",
       "      <th>whole_body</th>\n",
       "      <th>non_mosquito_Diptera_host</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>283 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                      PRODUCT\n",
       "gtdb_species                   internal_id mosquito_tissue sample_source                     \n",
       "Achromobacter ruhlandii        MMO-104     whole_body      mosquito                        31\n",
       "                               MMO-123     whole_body      mosquito                        31\n",
       "                               MMO-124     whole_body      mosquito                        31\n",
       "                               MMO-144     whole_body      mosquito                        31\n",
       "                               MMO-82      whole_body      mosquito                        31\n",
       "...                                                                                       ...\n",
       "Stenotrophomonas maltophilia   MMO-2       whole_body      mosquito                         5\n",
       "                               MMO-6       whole_body      mosquito                         5\n",
       "Stenotrophomonas maltophilia_A USHLN265    midgut          mosquito                         4\n",
       "                               USMM131     midgut          mosquito                         4\n",
       "UBA7405 sp000755535            AS183       whole_body      non_mosquito_Diptera_host       43\n",
       "\n",
       "[283 rows x 1 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# group by \n",
    "''''\n",
    "Summarise the number of MosAIC virulence factors across samples, species, tissues, source\n",
    "'''\n",
    "MosAIC_VF_Summary_Tisssue = MosAIC_VF_select[['internal_id', 'gtdb_species',\n",
    "                                               'PRODUCT', 'mosquito_tissue',\n",
    "                                                'sample_source']].groupby(['gtdb_species', 'internal_id', 'mosquito_tissue','sample_source']).count()\n",
    "\n",
    "\n",
    "MosAIC_VF_Summary_Tisssue.to_csv('140923_MosAIC_VF_Summary_Tissue.csv')\n",
    "MosAIC_VF_Summary_Tisssue\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
